cache data for background session task, allows reconnecting with back…

thuyBox opened the pull request:
…ground session task through associateId

thuyBox committed:
cache data for background session task, allows reconnecting with background session task through associateId

boxcla commented:
Verified that @thuyBox has signed the CLA. Thanks for the pull request!

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXContentSDKErrors.h:
```diff
@@ -81,7 +81,8 @@ typedef NS_ENUM(NSUInteger, BOXContentSDKURLSessionCacheError) {
     BOXContentSDKURLSessionCacheErrorNoValidSessionForUserId = 50001,
     BOXContentSDKURLSessionCacheErrorNoValidUserIdOrAssociateId = 50002,
-    BOXContentSDKURLSessionCacheErrorNoValidBackgroundSessionIdOrSessionTaskId = 50003
+    BOXContentSDKURLSessionCacheErrorNoValidBackgroundSessionIdOrSessionTaskId = 50003,
```
*NoValid* -> *Invalid* for these constants.

thuyBox replied:
done

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h:
```diff
@@ -160,4 +173,10 @@
  */
 - (BOOL)deleteCachedInfoForUserId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)error;
 
+- (NSString *)destinationFilePathGivenBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId;
```
Minor: `destinationFilePath*For*BackgroundSessionId`

thuyBox replied:
done

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h:
```diff
@@ -160,4 +173,10 @@
 
+- (NSString *)destinationFilePathGivenBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId;
+
+- (NSData *)responseDataGivenBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId;
```
Minor: responseData*For*BackgroundSessionId.

thuyBox replied:
done

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.h:
```diff
@@ -160,4 +173,10 @@
+
+- (NSData *)responseDataGivenBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId;
+
+- (BOXURLBackgroundSessionIdAndSessionTaskId *)backgroundSessionIdAndSessionTaskIdGivenUserId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)error;
```
Minor: backgroundSessionIdAndSessionTaskId*For*UserId

thuyBox replied:
done

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:
```diff
@@ -39,8 +38,8 @@ - (id)initWithCacheRootDir:(NSString *)cacheRootDir
-                [[NSFileManager defaultManager] createDirectoryAtPath:cacheDir withIntermediateDirectories:YES attributes:nil error:&error];
-                BOXAssert(error != nil, @"Failed to create cacheDir %@ with error %@", cacheDir, error);
+                BOOL success = [[NSFileManager defaultManager] createDirectoryAtPath:cacheDir withIntermediateDirectories:YES attributes:nil error:&error];
+                BOXAssert(success, @"Failed to create cacheDir %@ with error %@", cacheDir, error);
```
What's the benefit of asserting here for developers and/or us? Should we just return nil if it silently fails?

thuyBox replied:
yeah, assert to terminate in development mode. We will retry to create this dir when building subdir structures later. Do you think we should return nil and allow error as an argument for init so the user is aware of the issue?

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:
```diff
@@ -189,6 +187,32 @@ - (BOXURLSessionTaskCachedInfo *)cachedInfoForUserId:(NSString *)userId associat
+    NSString *filePath = [self filePathForBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId type:BOXURLSessionTaskCacheFileTypeDestinationFilePath];
+
+    //decrypt data found at filePath
+    NSData *data = [self unencryptedDataAtFilePath:filePath];
```
What if this is a big data file? Do we want to offload this to a background queue?

irvingruan replied:
I'll say the same about the encryption/decryption methods below.

thuyBox replied:
the data we encrypt/decrypt would be very small since they contain metadata around the session tasks, and they will be called on the thread of BOXAPIOperation.

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:
```diff
@@ -189,6 +187,32 @@ - (BOXURLSessionTaskCachedInfo *)cachedInfoForUserId:(NSString *)userId associat
+    return [NSKeyedUnarchiver unarchiveObjectWithData:data];
+}
+
+- (NSData *)responseDataGivenBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId
```
Minor: responseData*For*BackgroundSessionId

thuyBox replied:
done

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:
```diff
@@ -265,6 +270,31 @@ - (BOOL)cacheBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(
+- (NSString *)filePathForBackgroundSessionId:(NSString *)backgroundSessionId sessionTaskId:(NSUInteger)sessionTaskId type:(BOXURLSessionTaskCacheFileType)type
+{
+    NSString *path = [self dirPathOfSessionTaskWithBackgroundSessionId:backgroundSessionId sessionTaskId:sessionTaskId];
+    //persist data to sessions/$backgroundSessionId/$sessionTaskId/$fileType
```
Should we add this comment above this method call?

thuyBox replied:
done

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:
```diff
@@ -358,6 +388,8 @@ - (BOXURLBackgroundSessionIdAndSessionTaskId *)parseBackgroundSessionIdAndSessio
 
 - (BOXURLBackgroundSessionIdAndSessionTaskId *)backgroundSessionIdAndSessionTaskIdGivenUserId:(NSString *)userId associateId:(NSString *)associateId error:(NSError **)error
 {
+    if (userId == nil || associateId == nil) return nil;
```
Let's wrap this braces and on a different line.

thuyBox replied:
done

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionCacheClient.m:
```diff
@@ -358,6 +388,8 @@ - (BOXURLBackgroundSessionIdAndSessionTaskId *)parseBackgroundSessionIdAndSessio
+    if (userId == nil || associateId == nil) return nil;
+
     NSString *dir = [self dirPathOfUserSessionTaskGivenUserId:userId associateId:associateId];
     NSError *err;
```
Let's assign err to nil here.

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionManager.h:
```diff
@@ -95,44 +97,87 @@ totalBytesExpectedToSend:(int64_t)totalBytesExpectedToSend;
- Create a NSURLSessionUploadTask which can be run in background
+ Retrieve a NSURLSessionDownloadTask given a resume data to be run in the background to download file into a destination file path.
+
+ @param resumeData      data to resume download session task from
```
What do you think about using resumableData instead of resumeData?

thuyBox replied:
uhm, this name is what Apple uses for their method cancelByProducingResumeData, and downloadTaskWithResumeData, I'm afraid changing it might make ones think it's different from what's used by those methods

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:
```diff
@@ -115,17 +164,49 @@ - (NSURLSession *)createBackgroundSession
+        dispatch_semaphore_signal(sema);
+    }];
+
+    dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
```
The DISPATCH_TIME_FOREVER feels unsafe. Can we set a timeout here and a retry mechanism?

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:
```diff
@@ -167,50 +251,144 @@ - (void)deassociateSessionId:(NSString *)sessionId sessionTaskId:(NSUInteger)ses
+    if (userId == nil || associateId == nil) {
+        return YES;
+    }
+    NSError *err = nil;
```
We got an error and err variable. Let's be specific with what kind of "err" is.

thuyBox replied:
done, renamed the input error as outError instead

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:
```diff
@@ -167,50 +251,144 @@ - (void)deassociateSessionId:(NSString *)sessionId sessionTaskId:(NSUInteger)ses
+
+    if (task == nil) {
+        backgroundSessionId = backgroundSessionIdentifier;
+        sessionTask = uploadFromFileURL == nil ? (request != nil ? [self.backgroundSession downloadTaskWithRequest:request] : [self.backgroundSession downloadTaskWithResumeData:resumeDownloadData]) : [self.backgroundSession uploadTaskWithRequest:request fromFile:uploadFromFileURL];
```
This is long and hard to parse. Can we split this to multiple lines?

thuyBox replied:
done

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:
```diff
@@ -167,50 +251,144 @@ - (void)deassociateSessionId:(NSString *)sessionId sessionTaskId:(NSUInteger)ses
+    //cache background task if have not
+    NSError *error = nil;
+    BOOL success = [self.cacheClient cacheUserId:userId associateId:associateId backgroundSessionId:backgroundSessionId sessionTaskId:task.taskIdentifier error:&error];
+    BOXAssert(success, @"failed to cache background task %@", error);
```
What do you think about bubbling up an error here instead of asserting? I wonder if it's still advantageous to give back the background task even if persisting it to disk somehow failed, but at least there's an error exposed?

thuyBox replied:
done, exposing error

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:
```diff
@@ -219,32 +397,86 @@ - (NSURLSessionUploadTask *)createNonBackgroundUploadTaskWithStreamedRequest:(NS
+        if (![fileManager fileExistsAtPath:destinationFilePath]) {
+            success = [fileManager moveItemAtURL:temporaryURL toURL:destURL error:&err];
+        } else {
+            success = [fileManager replaceItemAtURL:destURL
```
Minor: let's align these colons.

thuyBox replied:
done

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:
```diff
@@ -219,32 +397,86 @@ - (NSURLSessionUploadTask *)createNonBackgroundUploadTaskWithStreamedRequest:(NS
-
-        if ([downloadTaskDelegate respondsToSelector:@selector(downloadTask:didFinishDownloadingToURL:)]) {
-            [downloadTaskDelegate downloadTask:downloadTask didFinishDownloadingToURL:location];
+    //this method is only called by background download tasks
```
Minor: Proper caps and space this comment.

thuyBox replied:
done

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:
```diff
@@ -219,32 +397,86 @@ - (NSURLSessionUploadTask *)createNonBackgroundUploadTaskWithStreamedRequest:(NS
-        if ([downloadTaskDelegate respondsToSelector:@selector(downloadTask:didFinishDownloadingToURL:)]) {
-            [downloadTaskDelegate downloadTask:downloadTask didFinishDownloadingToURL:location];
+    //this method is only called by background download tasks
+    if ([downloadTask.response isKindOfClass:[NSHTTPURLResponse class]]) {
```
Will .response ever be something different than NSHTTPURLResponse?

thuyBox replied:
we don't expect it to be, and if it is, we won't handle it

thuyBox committed:
update per comments, changing method names and expose error from persisting cache info for bg tasks

thuyBox committed:
allow reconnection from other background sessions
minor fixes for upload
wait for background session to populate pending session tasks before providing session tasks

irvingruan commented on BoxContentSDK/BoxContentSDK/BOXContentSDKConstants.h:
```diff
@@ -444,20 +444,22 @@ extern NSString *const BOXAPIEnterpriseEventTypeItemUnsync;
     BOXURLSessionTaskCacheFileTypeResponseData = 3,
-    BOXURLSessionTaskCacheFileTypeError = 4
+    BOXURLSessionTaskCacheFileTypeError = 4,
+    BOXURLSessionTaskCacheFileTypeUserIdAndAssociateId = 5
```
Minor: do we want error to be the last one?

irvingruan commented on BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.h:
```diff
@@ -156,7 +155,12 @@ typedef void (^BOXSessionTaskReplacedBlock)(NSURLSessionTask *oldSessionTask, NS
+ * Useful to reconnect to the task upon app restarts, or to resume a previously
+ * cancelled background download
+ */
+@property (nonatomic, readwrite, strong) NSString *associateId;
```
Copy for strings.

thuyBox replied:
done

irvingruan commented on BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m:
```diff
@@ -290,13 +290,42 @@ - (void)start
+        if (cachedInfo.response != nil && error == nil) {
+            //get valid cached info for session task, finish this operation
+
+            [self sessionTask:self.sessionTask didFinishWithResponse:cachedInfo.response responseData:cachedInfo.responseData error:cachedInfo.error];
```
We should check if the delegate implements this method.

thuyBox replied:
it's a required delegate method

irvingruan replied:
Sounds good

irvingruan commented on BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m:
```diff
@@ -519,19 +546,25 @@ - (void)finishURLSessionTaskWithData:(NSData *)data response:(NSURLResponse *)re
 {
-    [self processResponse:response];
+    @synchronized (self) {
+        [self processResponse:response];
```
Should we protect against nil response here?

thuyBox replied:
added a FIXME to review if we need a check there. It's a bit more thought needed to make sure we treat nil as valid/invalid based on the type of operations

irvingruan commented on BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m:
```diff
@@ -519,19 +546,25 @@ - (void)finishURLSessionTaskWithData:(NSData *)data response:(NSURLResponse *)re
 }
 
-- (void)processIntermediateData:(NSData *)data
+- (void)sessionTask:(NSURLSessionTask *)sessionTask processIntermediateData:(NSData *)data
```
Same here.

thuyBox replied:
done

irvingruan commented on BoxContentSDK/BoxContentSDK/Operations/BOXAPIOperation.m:
```diff
@@ -519,19 +546,25 @@ - (void)finishURLSessionTaskWithData:(NSData *)data response:(NSURLResponse *)re
 {
-    [self.responseData appendData:data];
+    @synchronized (self) {
+        [self.responseData appendData:data];
```
Same here.

thuyBox replied:
done

irvingruan commented on BoxContentSDK/BoxContentSDK/Requests/BOXFileUploadNewVersionRequest.m:
```diff
@@ -15,6 +15,7 @@ @interface BOXFileUploadNewVersionRequest ()
 
 @property (nonatomic, readwrite, strong) NSString *localFilePath;
 @property (nonatomic, readwrite, strong) NSString *uploadMultipartCopyFilePath;
+@property (nonatomic, readwrite, strong) NSString *associateId;
```
copy for strings.

thuyBox replied:
done

irvingruan commented on BoxContentSDK/BoxContentSDK/Requests/BOXFileUploadRequest.m:
```diff
@@ -18,6 +18,7 @@ @interface BOXFileUploadRequest ()
 
 @property (nonatomic, readwrite, strong) NSString *localFilePath;
 @property (nonatomic, readwrite, strong) NSString *uploadMultipartCopyFilePath;
+@property (nonatomic, readwrite, strong) NSString *associateId;
```
Copy for strings.

thuyBox replied:
done

irvingruan commented on BoxContentSDKSampleApp/BoxContentSDKSampleApp/BOXSampleAppDelegate.m:
```diff
@@ -29,8 +29,12 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
     UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:authenticationController];
     self.window.rootViewController = navController;
     [self.window makeKeyAndVisible];
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
```
Any reason why we're delaying a call to this?

thuyBox replied:
removed

thuyBox committed:
minor fixes per comments

irvingruan commented:
+2
